### PR TITLE
Add TypeScript type define and Promise support.

### PR DESCRIPTION
**Pull request by septs**

### CODE REVIEW

**Comment by septs** [comment]
#96 #90 type information provided is wrong.
#90 `deasync.await` implements an unhandled reject.

**Comment by septs** [comment]
@abbr

**Comment by abbr** [comment]
Thanks for the pr. I am ok for the ts but reluctant to add promise and await because they conceptually conflict with deasync. What's the use case that requires deasync support promise?

**Comment by septs** [comment]
Simultaneously provide users with asynchronous and synchronous.
```javascript
async function execCommand() { /* ...code... */ }
function execCommandSync() { return deasync.await(execCommand()) }
```

**Comment by septs** [comment]
@abbr

**Comment by septs** [comment]
If you really don't want to accept `Promise` support. I will remove the `Promise` support.

**Comment by septs** [comment]
@abbr

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+interface DeAsync {
+    <R>(fn: (callback: Callback<R>) => void): () => R;
+    <T1, R>(fn: (arg1: T1, callback: Callback<R>) => void): (arg1: T1) => R;
+    <T1, T2, R>(fn: (arg1: T1, arg2: T2, callback: Callback<R>) => void): (arg1: T1, arg2: T2) => R;
+    <T1, T2, T3, R>(fn: (arg1: T1, arg2: T2, arg3: T3, callback: Callback<R>) => void): (arg1: T1, arg2: T2, arg3: T3) => R;
+    <T1, T2, T3, T4, R>(fn: (arg1: T1, arg2: T2, arg3: T3, arg4: T4, callback: Callback<R>) => void): (arg1: T1, arg2: T2, arg3: T3, arg4: T4) => R;
+    <T1, T2, T3, T4, T5, R>(fn: (arg1: T1, arg2: T2, arg3: T3, arg4: T4, arg5: T5, callback: Callback<R>) => void): (arg1: T1, arg2: T2, arg3: T3, arg4: T4, arg5: T5) => R;
+    <T1, T2, T3, T4, T5, T6, R>(fn: (arg1: T1, arg2: T2, arg3: T3, arg4: T4, arg5: T5, arg6: T6, callback: Callback<R>) => void): (arg1: T1, arg2: T2, arg3: T3, arg4: T4, arg5: T5, arg6: T6) => R;
+    (fn: Function): Function
```
Comment (abbr) [inline index.d.ts]: I am reluctant to accept this incomplete enumeration of function arguments declaration style. I would like to wait until Typescript comes up with a solution, such as Microsoft/TypeScript/issues/5453.

Comment (septs) [reply index.d.ts]: I want to be able to merge and provide type hints for current users.

File: index.js [hunk 2]
```diff
@@ -73,4 +73,19 @@
...
+			result, error;
+		promise.then(
+			function (value) { resolved = true; result = value; return result; },
+			function (reason) { rejected = true; error = reason; return reason; }
+		)
+		module.exports.loopWhile(function () { return !resolved && !rejected; })
+		if (rejected) {
+			throw error;
+		}
+		return result;
```
Comment (milahu) [inline index.js:89]: this does not work with `async terser.minify()`
the promise is never resolved or rejected = deadloop

`return result` and `return reason` in the callback fns is pointless

**Comment by FFdhorkin** [comment]
> Thanks for the pr. I am ok for the ts but reluctant to add promise and await because they conceptually conflict with deasync. What's the use case that requires deasync support promise?

In what way do they "conceptually conflict" ? The whole point of deasync is to be able to run async code synchronously, isn't it?

For example, suppose you have the following and you need to have to run it in a synchronous context:

```ts
async function foo() {
    return true;
}
```

My expectation was I'd be able to then do this in a place where I cannot make a method asynchronous in order to use await:

```ts
const syncFoo = deasync(foo);
let result = syncFoo();
somethingElse(result); // won't run until the promise from foo resolves
```

But this gives various type errors. I haven't actually tried _running_, but presumably if TypeScript is complaining about it, then `deasync` as a library can't currently do the one use case I actually need it for.

This is shocking to me, given the library is described as `DeAsync turns async function into sync`. Async functions **_by definition_** return promises... The example with `child_process.exec` might be _asynchronous process creation_, but it is **not** an async function; it immediately returns the `ChildProcess` object!

**Comment by milahu** [comment]
> I haven't actually tried _running_

just try, for science ; )

alternatives to deasync: https://github.com/terser/terser/issues/801#issuecomment-686754871

**Pull request closed without merging**